### PR TITLE
#15123 fix performance unnecessary value param

### PR DESCRIPTION
**Pull request by afuller-TT**
### Ticket
#15123 

### Problem description
Several functions force a copy of non-trivial types when a const ref would suffice.

### What's changed
* Converted to const-ref where suitable.  Done via automated FIX-ITs in clang-tidy. Commit 59d44d8
* Manually fixed the build failure that was a result of not having a header to share the declaration of some functions.  Commit 657d064
  * A better fix is probably to use a header, but this PR is already sprawling, I opted to keep the changes tightly focused.
* Manually fix the diagnostics that clang-tidy didn't attempt to auto correct.  It appears the FIX-ITs aren't applied to template functions or lambdas. Commit 002c902

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12024045658
- [x] Blackhole Post commit (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/12036802773
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: .clang-tidy
Comment (afuller-TT) [inline .clang-tidy:1]: You may wish to review commit-by-commit.

The first is Clang Tidy's automatic code changes and is the biggest diff.
The two following are my manual edits.  Smaller diff, but with human involvement.

**Review by pavlejosipovic — APPROVED** [review]
Checked conv2d. 
Thank you!

File: tests/tt_metal/tt_metal/perf_microbenchmark/old/matmul/matmul_local_l1.cpp [hunk 1]
```diff
@@ -158,7 +158,7 @@ std::vector<T> slice(std::vector<T> const &v, int m, int n) {
   return vec;
 }
 
-void print_vec(std::vector<bfloat16> data, int rows, int cols, string name) {
+void print_vec(std::vector<bfloat16> data, int rows, int cols, const string& name) {
```
Comment (ayerofieiev-tt) [inline tests/tt_metal/tt_metal/perf_microbenchmark/old/matmul/matmul_local_l1.cpp:161]: const ref? :)

**Review by mo-tenstorrent — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]
Lets go!

**Review by bbradelTT — APPROVED** [review]
If this provides a benefit I'm for it.

Could you please elaborate on the benefit of `std::optional<>&` vs `std::optional>?`?

I thought optional was supposed to reasonably efficient.

**Comment by afuller-TT** [comment]
> If this provides a benefit I'm for it.
> 
> Could you please elaborate on the benefit of `std::optional<>&` vs `std::optional>?`?
> 
> I thought optional was supposed to reasonably efficient.

Thanks for the nudge.  I'm sure optional is as efficient as it is able to be and still be capable of doing its job.  And unlikely that these calls are bottlenecks in our performance.  But when we can be more efficient with an extra char or six and a tool will even tell us when and where to place, them, why leave efficiency on the table? :)

_How_ much impact it has will come down to how large the object is or how complex the copy c'tor is.  Someone on S.O. wrote a nice commentary with graphs: https://stackoverflow.com/a/77161349

**Review by dmakoviichuk-tt — APPROVED** [review]

**Review by rfurko-tt — APPROVED** [review]
Great work, thanks for doing this!
Do you have an estimate for performance impact for the PR? Thanks in advance!

**Review by SeanNijjar — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/permute/device/permute_program_factory.cpp [hunk 2]
```diff
@@ -8,17 +8,17 @@
 namespace ttnn::operations::data_movement {
 
 namespace detail {
-    uint32_t num_pages(const ttnn::Tensor& input_tensor) {
+    uint32_t num_pages(const ttnn::Tensor input_tensor) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/permute/device/permute_program_factory.cpp]: This causes a copy

**Review by cfjchu — APPROVED** [review]

**Comment by afuller-TT** [comment]
Post commit and Blackhole are green.  Merging this in.  I'll check main this afternoon in case new code fails the newly activated clang-tidy check.

**Pull request merged**